### PR TITLE
Address cholesky issue with arrowhead mass trix

### PR DESCRIPTION
**Pull request by fehiepsi**
This has seen to be helpful in @martinjankowiak 's sir experiments, so I guess we can include it in the release. The idea is to scale non-diagonal terms down if Cholesky issue happens.

### CODE REVIEW

File: pyro/ops/arrowhead.py [hunk 1]
```diff
@@ -28,12 +28,22 @@ def sqrt(x):
...
     # ref: https://en.wikipedia.org/wiki/Schur_complement#Background
     Dsqrt = x.bottom_diag.sqrt()
-    B_Dsqrt = B / Dsqrt.unsqueeze(-2)  # shape: head_size x N
-    schur_complement = A - B_Dsqrt.matmul(B_Dsqrt.t())  # complexity: head_size^2 x N
-    # we will decompose schur_complement to U @ U.T (so that the sqrt matrix
-    # is upper triangular) using some `flip` operators:
-    #   flip(cholesky(flip(schur_complement)))
-    top_left = torch.flip(torch.flip(schur_complement, (-2, -1)).cholesky(), (-2, -1))
+
+    for i in range(6):
```
Comment (fritzo) [inline pyro/ops/arrowhead.py:32]: nit: could you add a comment like
```py
# On cholesky error, retry with smaller tail part B.
num_attempts = 6
for i in range(num_attempts):
    ...
```

File: pyro/ops/arrowhead.py [hunk 2]
```diff
@@ -28,12 +28,22 @@ def sqrt(x):
...
+        B_Dsqrt = B / Dsqrt.unsqueeze(-2)  # shape: head_size x N
+        schur_complement = A - B_Dsqrt.matmul(B_Dsqrt.t())  # complexity: head_size^2 x N
+        # we will decompose schur_complement to U @ U.T (so that the sqrt matrix
+        # is upper triangular) using some `flip` operators:
+        #   flip(cholesky(flip(schur_complement)))
+        try:
+            top_left = torch.flip(torch.cholesky(torch.flip(schur_complement, (-2, -1))), (-2, -1))
+            break
+        except RuntimeError:
+            B = B * 2 ** (-i - 1)
```
Comment (fritzo) [inline pyro/ops/arrowhead.py:42]: Can you confirm that you want super-exponential growth here, not just exponential growth via `B = B / 2`?

Comment (fritzo) [reply pyro/ops/arrowhead.py:42]: @martinjankowiak can you confirm this super-exponential backoff is the desired behavior?

Comment (martinjankowiak) [reply pyro/ops/arrowhead.py:42]: i can confirm it's what i previously did experiments with

https://github.com/pyro-ppl/pyro/blob/2c709199d5932d48e1d409a9275a4ff67f6e6549/pyro/ops/arrowhead.py#L47

it's probably not the platonic solution but i guess it doesn't hurt... i suspect that in most cases where arrowhead is running into troubles here simplistic regularization will be moderately helpful at best

File: pyro/ops/arrowhead.py [hunk 3]
```diff
@@ -28,12 +28,22 @@ def sqrt(x):
...
+            continue
+        raise RuntimeError("Singular schur complement in computing Cholesky of the input"
+                           " arrayhead matrix")
```
Comment (martinjankowiak) [inline pyro/ops/arrowhead.py:45]: typo: arrayhead

File: pyro/ops/arrowhead.py (hunk 2 above)
Comment (fehiepsi) [reply pyro/ops/arrowhead.py:42]: Thanks for reviewing, Fritz! It is not the desired behavior. I will make the change.

**Pull request merged**